### ISSUE TITLE
Assertion failed: (tiles.find(key) == tiles.end())

### ISSUE THREAD

**Issue by jfirebaugh** [mapbox/mapbox-gl-native#1308]
I get this assertion failure pretty frequently when panning and zooming in the desktop app:

```
Assertion failed: (tiles.find(key) == tiles.end()), function add, file /Users/john/Development/mapbox-gl-native/src/mbgl/map/tile_cache.cpp, line 23.
```

**Comment by 1ec5** [comment]
Seems to happen more easily if you start out at z1 or so then zoom in.

**Comment by incanus** [comment]
The easy — and very tempting — answer here is to check before doing a cache insertion. But that's a performance hit and I figured @ansis had some assumptions that would make this unnecessary. 

I went through a number (more than six!) hypotheses about what was really going on here, but it actually comes down to **world wrap**. It's why you don't see it until low zooms are involved (or you pan a **lot** to wrap the world at higher zooms, which I'm guessing no one's doing or is doing infrequently — I wasn't able to do it). It's also why you don't see it much on taller-than-wide iOS devices but do on wider-than-tall OS X. 

We aren't normalizing the tiles inserted into cache, but we **do** normalize searches in `tile_data` inside of `addTile()`. Here's a trigger scenario: 

- Pan so both `1/-1/0` and `1/1/0` are in the viewport. Both share the normalized `1/1/0` key in `tile_data` and don't yet exist in cache. 

- Zoom in to `z2` and `1/1/0` is considered a candidate parent for coverage. This keeps it in `retain` and thus `tile_data`. 

- Pan around or zoom in further, loading more native-zoom tiles, enough that `1/1/0` becomes obsolete and is moved out of `tile_data` into cache as `key = 33` (the `uint64_t` version). 

- _Unrelated, but also needs fixed:_ It doesn't matter what happens to `1/-1/0`, as it gets inserted into cache as `key = <a little less than MAX_INT>` as cache tries to convert it to a `uint64_t` and it underflows due to its negative `x` value. 

- Zoom back out to `z1` such that both `1/-1/0` and `1/1/0` are visible. Since they load sequentially, a situation could occur where the load of `1/-1/0` happens first, normalizes to `1/1/0`, and gets added to `tile_data` as such. When it gets to `1/1/0`, it's considered already in `tile_data` and doesn't query — and thus doesn't modify — cache. 

- Pan or zoom such that `1/1/0` is obsolete and attempts to add to cache again as `key = 33`. 

- :bomb::fire::exclamation: 

The solution is to normalize cache lookups and adds just like `addTile()` does.
